tutorials/mnist_tutorial_pure_tf.py

```diff
@@ -31,32 +32,56 @@
 """
 
 
-class MLP(object):
+class MLP(Model):
   """
   An example of a bare bones multilayer perceptron (MLP) class.
   """
 
   def __init__(self, layers, input_shape):
+    super(MLP, self).__init__()
+
+    self.layer_names = []
     self.layers = layers
     self.input_shape = input_shape
-    for layer in self.layers:
+    for i,layer in enumerate(self.layers):
+      name = layer.__class__.__name__
+      self.layer_names.append(name+"_"+str(i))
+
       layer.set_input_shape(input_shape)
       input_shape = layer.get_output_shape()
 
-  def fprop(self, x, return_all=False, set_ref=False):
+  def get_layer_names(self):
+    return self.layer_names
+
+  def fprop(self, x, set_ref=False):
+    if (x, self.state) in self.fprop_cache.keys():
```

Comment:
OK, this is concerning. I don't think everyone who implements a symbolic fprop should have to implement a caching mechanism. I thought we were going to put a caching mechanism in a base class's generate_np and then no one else would need to implement a new caching mechanism.